File: CustomDrawing/CustomDrawing/ViewController.m
//
//  ViewController.m
//  CustomDrawing
//
//  Created by Aoife McManus on 9/29/21.
//

#import "ViewController.h"
#import <UIKit/UIGraphicsRendererSubclass.h>
#import <CoreGraphics/CGPath.h>

@interface ViewController ()

@end

@implementation ViewController
@synthesize YourView;

struct Spirograph{
    int innerR;
    int outerR;
    int distance;
    CGFloat amount;
};

- (int)getGCD:(int)innerR
joiningArgument2:(int)outerR{
    if(innerR==outerR){
        return innerR;
    }
    if(innerR>outerR){
        return [self getGCD:innerR-outerR joiningArgument2:outerR];
    }
    return [self getGCD:innerR joiningArgument2:outerR-innerR];
}

- (CGMutablePathRef)getPath:(struct Spirograph)spirograph{
    int div=[self getGCD:spirograph.innerR joiningArgument2:spirograph.outerR];
    CGFloat outerR=(CGFloat)spirograph.outerR;
    CGFloat innerR=(CGFloat)spirograph.innerR;
    CGFloat distance=(CGFloat)spirograph.distance;
    CGFloat difference=innerR-outerR;
    CGFloat end=ceilf(2*M_PI*outerR/(CGFloat)div*spirograph.amount);
    CGMutablePathRef ref=CGPathCreateMutable();
    for(float i=0; i<end; i+=0.01){
        CGFloat x=difference*cos(i)+distance*cos(difference/outerR*i);
        CGFloat y=difference*sin(i)+distance*sin(difference/outerR*i);
        if(i==0) CGPathMoveToPoint(ref, NULL, x, y);
        else{
            CGPathAddLineToPoint(ref, NULL, x, y);
        }
    }
    return ref;
}

- (UIImageView*)getView:(CGMutablePathRef)path{
    CAShapeLayer* shapeLayer=[[CAShapeLayer alloc]init];
    shapeLayer.path=path;
    UIColor* purp=[UIColor colorWithRed: 0.69 green: 0.41 blue: 0.94 alpha: 1.00];
    UIColor* white=[UIColor colorWithRed: 1 green: 1 blue: 1 alpha: 1.00];

    shapeLayer.strokeColor=purp.CGColor;
    shapeLayer.fillColor=white.CGColor;

    shapeLayer.lineWidth=1;
    //shapeLayer.position=CGPointMake(10, 10);
    UIImageView* view=[[UIImageView alloc]initWithFrame:CGRectMake(210, 400, 300, 300)];
    [view.layer insertSublayer:shapeLayer atIndex:0];
    return view;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    /*UIImage* returnImage=[self getImage];
    UIImageView *dot =[[UIImageView alloc] initWithImage:returnImage];
    [self.view addSubview:dot];*/
    struct Spirograph spirograph;
    spirograph.innerR=75.0;
    spirograph.outerR=125.0;
    spirograph.distance=25.0;
    spirograph.amount=1.0;
    CGMutablePathRef ref=[self getPath:spirograph];
    UIImageView* subView=[self getView:ref];
    [self.view addSubview:subView];
    [UIView animateWithDuration:0.5f animations:^{subView.frame = CGRectOffset(subView.frame, 0, 50);
    }];
    
}

- (UIImage*)getImage{
    UIGraphicsImageRenderer *renderer = [[UIGraphicsImageRenderer alloc] initWithSize:CGSizeMake(200, 200)];
    UIImage *image = [renderer imageWithActions:^(UIGraphicsImageRendererContext * _Nonnull context) {
        [[UIColor darkGrayColor] setStroke];
        [context strokeRect:renderer.format.bounds];
        [[UIColor colorWithRed:158/255.0 green:215/255.0 blue:245/255.0 alpha:1] setFill];
        [context fillRect:CGRectMake(1, 1, 140, 140)];
        [[UIColor colorWithRed:145/255.0 green:211/255.0 blue:205/255.0 alpha:1] setFill];
        [context fillRect:CGRectMake(60, 60, 140, 140) blendMode:kCGBlendModeMultiply];
        
        [[UIColor colorWithRed:203/255.0 green:222/255.0 blue:116/255.0 alpha:0.6] setFill];
        CGContextFillEllipseInRect(context.CGContext, CGRectMake(60, 60, 140, 140));
      }];
    return image;
}



@end

File: CustomDrawing/CustomDrawing/ViewController.swift
//
//  ViewController.swift
//  CustomDrawing
//
//  Created by Aoife McManus on 10/4/21.
//

import UIKit

class ViewController: UIViewController {
    
    struct Spirograph{
        var innerR: Int
        var outerR: Int
        var distance: Int
        var amount: CGFloat
    }
    
    func getGCD(innerR: Int, outerR: Int) -> Int{
        if(innerR==0){
            return outerR
        }
        if(outerR==0){
            return innerR
        }
        if(innerR==outerR){
            return innerR
        }
        if(innerR>outerR){
            return getGCD(innerR:innerR-outerR, outerR:outerR)
        }
        return getGCD(innerR:innerR, outerR:outerR-innerR)
    }
    
    func getPath(spirograph: Spirograph)->UIBezierPath{
        let div=getGCD(innerR: spirograph.innerR, outerR:spirograph.outerR)
        let outerR=CGFloat(spirograph.outerR)
        let innerR=CGFloat(spirograph.innerR)
        let distance=CGFloat(spirograph.distance)
        let difference=innerR-outerR
        let end=ceil(2*CGFloat.pi*CGFloat(outerR)/CGFloat(div))*spirograph.amount
        var path:UIBezierPath!
        path=UIBezierPath()
        for i in stride(from:0, to:end, by:0.01){
            let x=difference*cos(i)+distance*cos(difference/outerR*i);
            let y=difference*sin(i)+distance*sin(difference/outerR*i);
            if(i==0){
                path.move(to: CGPoint(x:x, y:y))
            }
            else{
                path.addLine(to: CGPoint(x:x, y:y))
            }
        }
        return path
    }
    
    
    weak var shapeLayer: CAShapeLayer?

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.shapeLayer?.removeFromSuperlayer()
        let spiro=Spirograph(innerR: 75, outerR:125, distance:25, amount:1.0)
        let path=getPath(spirograph: spiro)
        

        // create shape layer for that path
        let width=CGFloat(125);
        let height=CGFloat(125);
        let x=self.view.frame.midX;
        let y=self.view.frame.midY;
        
        let shapeLayer = CAShapeLayer()
        let rect=CGRect(x: x, y: y, width: width, height: height)
        
        shapeLayer.frame = rect;
        shapeLayer.fillColor = #colorLiteral(red: 0, green: 0, blue: 0, alpha: 0).cgColor
        shapeLayer.strokeColor = #colorLiteral(red: 0.5568627715, green: 0.3529411852, blue: 0.9686274529, alpha: 1).cgColor
        shapeLayer.lineWidth = 2
        shapeLayer.path = path.cgPath

        // animate it

        view.layer.addSublayer(shapeLayer)
        let animation = CABasicAnimation(keyPath: "strokeEnd")
        animation.fromValue = 0
        animation.duration = 5;
        shapeLayer.add(animation, forKey: "MyAnimation")

        // save shape layer

        self.shapeLayer = shapeLayer
    }
}

